lib/ProductOpener/APIProductWrite.pm

```diff
@@ -0,0 +1,210 @@
+					my $packaging_ref
+						= get_checked_and_taxonomized_packaging_component_data($request_body_ref->{tags_lc},
+						$input_packaging_ref, $response_ref);
+					# Add or combine with the existing packagings components array
+					add_or_combine_packaging_component_data($product_ref, $packaging_ref, $response_ref);
+				}
+			}
+		}
+	}
+	return;
+}
+
+=head2 write_product_api()
+
+Process API v3 WRITE product requests.
+
+=cut
+
+sub write_product_api ($request_ref) {
+
+	$log->debug("write_product_api - start", {request => $request_ref}) if $log->is_debug();
+
+	my $response_ref = $request_ref->{api_response};
+
+	decode_json_request_body($request_ref);
+	my $request_body_ref = $request_ref->{request_body_json};
+
+	$log->debug("write_product_api - body", {request_body => $request_body_ref}) if $log->is_debug();
+
+	if (not defined $request_body_ref) {
+		$log->error("write_product_api - missing or invalid input body", {}) if $log->is_error();
+	}
+	elsif (not defined $request_body_ref->{product}) {
+		$log->error("write_product_api - missing input product", {request_body => $request_body_ref})
+			if $log->is_error();
+		add_error(
+			$response_ref,
+			{
+				message => {id => "missing_field"},
+				field => {id => "product"},
+				impact => {id => "request_failed"},
+			}
+		);
+	}
+	else {
+
+		# Use default request language if we did not get tags_lc
+		if (not defined $request_body_ref->{tags_lc}) {
+			$request_body_ref->{tags_lc} = $lc;
+			add_warning(
+				$response_ref,
+				{
+					message => {id => "missing_field"},
+					field => {id => "tags_lc", default_value => $request_body_ref->{tags_lc}},
+					impact => {id => "warning"},
+				}
+			);
+		}
+
+		# Load the product
+		my $code = $request_ref->{code};
+		my $product_id = product_id_for_owner($Owner_id, $code);
+		my $product_ref = retrieve_product($product_id);
+
+		if (not defined $product_ref) {
+			$product_ref = init_product($User_id, $Org_id, $code, $country);
+			$product_ref->{interface_version_created} = "20221102/api/v3";
+		}
+
+		# Process edit rules
+
+		$log->debug("phase 0 - checking edit rules", {code => $code}) if $log->is_debug();
+
+		my $proceed_with_edit = process_product_edit_rules($product_ref);
+
+		$log->debug("phase 0", {code => $code, proceed_with_edit => $proceed_with_edit}) if $log->is_debug();
+
+		if (not $proceed_with_edit) {
+
+			add_error(
+				$response_ref,
+				{
+					message => {id => "edit_against_edit_rules"},
+					field => {id => "product"},
+					impact => {id => "request_failed"},
+				}
+			);
+		}
+		else {
+
+			# Update the product
+			update_product_fields($request_ref, $product_ref);
+
+			# Process the product data
+			analyze_and_enrich_product_data($product_ref);
+
+			# Save the product
+			my $comment = $request_body_ref->{comment} || "API v3";
+			if (store_product($User_id, $product_ref, $comment)) {
+				# Notify robotoff
+				send_notification_for_product_change($product_ref, "updated");
+			}
```

Comment:
we should move that in store_product like for off-search indexation, it's silly to repeat this notification everywhere…

Reply:
ok